### ISSUE TITLE
[LSP Diagnostics Feature Enhancement] Add severity_limit to on_publish_diagnostics (if possible)

### ISSUE THREAD

**Issue by bluz71** [neovim/neovim#13314]
Hello Neovim LSP folks,

The new `vim.lsp.with` + `on_publish_diagnostics` API for tweaking LSP diagnostics is very nice. Well done.

However, I still find the *TypeScript Language Server* way too chatty with regards to issues it reports.
Neovim LSP diagnostics classifies issues into four categories: Error, Warning, Information and Hint. The *TypeScript Language Server* is showing a bunch of hints which really aren't issues, nevertheless it is clutter that I want to be rid of (at a configuration level).

It would be nice to be able to tweak Neovim LSP Diagnostics to only display issues above a certain severity limit. The function `get_line_diagnostics` already is aware of the concept of `severity_limit`, hence, it would be nice to extend that (if possible) also to `vim.lsp.with`.

For example:

```viml
vim.lsp.handlers["textDocument/publishDiagnostics"] = vim.lsp.with(
  vim.lsp.diagnostic.on_publish_diagnostics, {
    signs = true,
    underline = false,
    update_in_insert = false,
    virtual_text = {
      spacing = 2,
    },
    severity_limit = 'Warning' <<< NEW OPTION, MAYBE
  }
)
```

This theoretical example states only show errors and warning, not information and hints.

Best regards.

**Comment by glepnir** [comment]
It’s easy to implement, but I’m thinking whether #13295 can do this? @tjdevries

**Comment by windwp** [comment]
lol i agree with @bluz71  tsserver throw too many issue on warning.
I can accept to view it on sign but the virtual text is so annoy. it will better if we can do this

``` vim
vim.lsp.handlers["textDocument/publishDiagnostics"] = vim.lsp.with(
  vim.lsp.diagnostic.on_publish_diagnostics, {
    signs = {
     severity_limit='all'
     },
    underline = false,
    update_in_insert = false,
    virtual_text = {
      spacing = 2,
     severity_limit='Error'  ==> Only show virtual text on error
    },
  }
)
```

**Comment by bluz71** [comment]
In my case I desire severity_limit for both virtual_text **and** signs.

#13378 appears to only target virtual_text. Can signs also be added please.

Many thanks.

**Comment by glepnir** [comment]
yep. I had add it in pr.

**Comment by tjdevries** [comment]
Can any of you test this PR https://github.com/neovim/neovim/pull/13528 and make sure it works how you're expecting.

**Comment by bluz71** [comment]
Hello TJ,

I just tested out #13528 with this configuration:

```lua
local diagnostic_handler = vim.lsp.with(
  vim.lsp.diagnostic.on_publish_diagnostics, {
    signs = {
      severity_limit = 'Warning',
    },
    underline = false,
    update_in_insert = false,
    virtual_text = {
      spacing = 2,
      severity_limit = 'Warning',
    },
  }
)
```

Severity limit makes most sense when applied to both signs and virtual text (as listed above).

The PR works very well from my testing. Two big thumbs up from me. I look forward to the merge.

P.S. Side note whilst I am here, the one last piece of the puzzle that is preventing me from switching from ALE to native Neovim LSP is that diagnostics are always emitted when opening a buffer with Neovim LSP. In ALE land one can delay diagnostic display until file save which I find hugely beneficial. Issue #13324 already, just giving that a plug here (I am probably speaking out of turn).

P.S.S. I do genuinely appreciate all the great work the Neovim contributors have made. Fixing this issue here is a great step forward. Best regards.

**Comment by windwp** [comment]
+1 It's is work for me
Thank

**Comment by tjdevries** [comment]
Closed by #13528 

Thanks for confirming!
